### PR TITLE
allow tokens and hotwords be loaded from buffered string driectly

### PR DESCRIPTION
**Pull request by shawl336**
for online recognizer sherpa-onnx models, allow tokens and hotwords be loaded from buffered string driectly

### CODE REVIEW

File: sherpa-onnx/csrc/online-model-config.h [hunk 1]
```diff
@@ -69,7 +73,8 @@ struct OnlineModelConfig {
         debug(debug),
         model_type(model_type),
         modeling_unit(modeling_unit),
-        bpe_vocab(bpe_vocab) {}
+        bpe_vocab(bpe_vocab),
+        tokens_buf_str("") {}
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-model-config.h]:
```suggestion
        bpe_vocab(bpe_vocab) {}
```

`std::string` is default intialized to empty, so there is no need
to re-initialize it here.

File: sherpa-onnx/csrc/online-model-config.cc [hunk 2]
```diff
@@ -56,7 +56,7 @@ bool OnlineModelConfig::Validate() const {
     return false;
   }
 
-  if (!FileExists(tokens)) {
+  if (tokens_buf_str.empty() && !FileExists(tokens)) {
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-model-config.cc]: Please validate that either tokens is empty or tokens_buf_str is empty.

The cannot be non-empty or empty at the same time.

File: sherpa-onnx/csrc/online-model-config.h [hunk 3]
```diff
@@ -45,6 +45,10 @@ struct OnlineModelConfig {
   std::string modeling_unit = "cjkchar";
   std::string bpe_vocab;
 
+  /// if tokens_buf_str is non-empty, 
+  /// the tokens will be loaded from the buffered string in prior to the ${tokens} file
+  std::string tokens_buf_str;
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-model-config.h]:
```suggestion
  std::string tokens_buf;
```

`tokens_buf` is shorter.

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h [hunk 4]
```diff
@@ -83,7 +83,8 @@ class OnlineRecognizerTransducerImpl : public OnlineRecognizerImpl {
       : OnlineRecognizerImpl(config),
         config_(config),
         model_(OnlineTransducerModel::Create(config.model_config)),
-        sym_(config.model_config.tokens),
+        sym_(config.model_config.tokens_buf_str.empty() ? config.model_config.tokens :
+             config.model_config.tokens_buf_str, config.model_config.tokens_buf_str.empty() ? true : false),
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-recognizer-transducer-impl.h]: Line 86-87 are hard to read.

Could you put the initialization inside `{}`, not in the initializer list?

File: sherpa-onnx/c-api/c-api.h [hunk 5]
```diff
@@ -147,6 +149,9 @@ SHERPA_ONNX_API typedef struct SherpaOnnxOnlineRecognizerConfig {
   const char *rule_fsts;
   const char *rule_fars;
   float blank_penalty;
+  
+  /// if non-nullptr, loading the hotwords from the buffered string directly in
+  const char *hotwords_buf_str;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h]: What if there is a `0` inside `hotwords_buf_str`?

Please also add a member
```
int32_t hotwords_buf_str_len;
```

File: c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c [hunk 6]
```diff
@@ -0,0 +1,655 @@
+// c-api-examples/streaming-zipformer-c-api.c
+//
+// Copyright (c)  2024  Xiaomi Corporation
```
Comment (csukuangfj) [inline c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c:3]: Please add your info here.

File: c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c [hunk 7]
```diff
@@ -0,0 +1,655 @@
+// c-api-examples/streaming-zipformer-c-api.c
```
Comment (csukuangfj) [inline c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c]: Please update this line to match the actual file.

File: c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+  SherpaOnnxDestroyOnlineRecognizer(recognizer);
+  fprintf(stderr, "\n");
+
+  return 0;
+}
+
+const char* hotwords_buf_str = "▁A ▁T ▁P :1.5\n \
+▁A ▁B ▁C :3.0";
+
+const char* tokens_buf_str = "<blk> 0 \n \
```
Comment (csukuangfj) [inline c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c]: Could you read it from `tokens.txt`?

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c (hunk 8 above)
Comment (shawl336) [reply c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c]: do you mean having an external tokens.txt file,  reading its content as a string and pass the string to the api?

Comment (csukuangfj) [reply c-api-examples/streaming-zipformer-buffered-tokens-hotwords-c-api.c]: Yes, you are right. In that case, users can easily change the demo code to use another `tokens.txt`.

**Pull request closed without merging**